charmcraft/commands/store/registry.py

```diff
@@ -440,3 +448,80 @@ def _upload_blob(self, filepath, size, digest):
 
         # finally remove the temp filepath
         os.unlink(filepath)
+
+    def upload_from_local(self, digest: str) -> str:
+        """Upload the image from the local registry."""
+        dockerd = LocalDockerdInterface()
+
+        # validate the image is present locally
+        logger.debug("Checking image is present locally")
+        image_info = dockerd.get_image_info(digest)
+        if image_info is None:
+            return
+        local_image_size = image_info["Size"]
+
+        logger.debug("Getting the image from the local repo; size=%s", local_image_size)
+        response = dockerd.get_streamed_image_content(digest)
+
+        tmp_exported = tempfile.NamedTemporaryFile(mode="wb", delete=False)
+        extracted_total = 0
+        for chunk in response.iter_content(CHUNK_SIZE):
+            extracted_total += len(chunk)
+            progress = 100 * extracted_total / local_image_size
+            print("Extracting... {:.2f}%\r".format(progress), end="", flush=True)
+            tmp_exported.file.write(chunk)
+        tmp_exported.close()
+
+        # open the image tar and inspect it to get the config and layers from the only one
+        # manifest inside (as it's a list of one)
+        image_tar = tarfile.open(tmp_exported.name)
+        local_manifest = json.load(image_tar.extractfile("manifest.json"))
+        (local_manifest,) = local_manifest
+        config_name = local_manifest.get("Config")
+        layer_names = local_manifest["Layers"]
+        manifest = {
+            "mediaType": MANIFEST_V2_MIMETYPE,
+            "schemaVersion": 2,
+        }
+
+        if config_name is not None:
+            fpath, size, digest = self._extract_file(image_tar, config_name)
+            logger.debug("Uploading config blob, size=%s, digest=%s", size, digest)
+            self._upload_blob(fpath, size, digest)
+            manifest["config"] = {
+                "digest": digest,
+                "mediaType": CONFIG_MIMETYPE,
+                "size": size,
+            }
+
+        manifest["layers"] = manifest_layers = []
+        for idx, layer_name in enumerate(layer_names, 1):
+            fpath, size, digest = self._extract_file(
+                image_tar, layer_name, compress=True
+            )
+            logger.debug(
+                "Uploading layer blob %s/%s, size=%s, digest=%s",
+                idx,
+                len(layer_names),
+                size,
+                digest,
+            )
+            self._upload_blob(fpath, size, digest)
+            manifest_layers.append(
+                {
+                    "digest": digest,
+                    "mediaType": LAYER_MIMETYPE,
+                    "size": size,
+                }
+            )
+
+        # remove the temp tar file
+        os.unlink(tmp_exported.name)
```

Comment:
Do we need to guarantee that this file will be removed if the upload fails?

Reply:
As we just discussed, this is good enough. May be improved in the future if really necessary, when we see the value of the added complexity.